NtGetNextThread exception in UWP apps

For some reason in Minecraft `NtGetNextThread` throws an exception with code `0xC0000008: An invalid handle was specified.`, causing game crash, this is a UWP game afaik but I did use some launcher to set it up to let me copy DLLs into it.

IIRC some APIs are restricted in UWP land, guess maybe NtGetNextThread is one of them? Never had this exception in any non-UWP apps at least.

Call stack:
```C
 	ntdll.dll!KiRaiseUserExceptionDispatcher()	Unknown
>	dxgi.dll!safetyhook::ThreadFreezer::ThreadFreezer(void)	C++
 	dxgi.dll!safetyhook::InlineHook::e9_hook(class std::shared_ptr<class safetyhook::Allocator> const &)	C++
 	dxgi.dll!safetyhook::InlineHook::create(class std::shared_ptr<class safetyhook::Allocator> const &,unsigned __int64,unsigned __int64)	C++
 	dxgi.dll!safetyhook::InlineHook::create(unsigned __int64,unsigned __int64)	C++
 	dxgi.dll!safetyhook::create_inline(void *,void *)	C++
 	dxgi.dll!nvngx::init(struct HINSTANCE__ *)	C++
```

FWIW the older toolhelp method did used to work fine on this game too.

If you own Minecraft Bedrock (or have Xbox Game Pass PC) you can use https://github.com/bedrockLauncher/BedrockLauncher to set the game up in a writable folder, then build a wrapper DLL to load into it and try creating an inline hook.

E: strange, if I run with debugger attached and manually continue past the exception it runs fine, seems that some of the calls to NtGetNextThread are also successful too, not sure why this is causing a crash when debugger isn't attached...
(also disabling the exception type in VS debugger is still making it prompt for each exception too, not sure if that means anything)

Seems wrapping the NtGetNextThread call in a SEH handler works to stop it crashing the whole app:

```C
// NtGetNextThread may throw a SEH exception that causes some apps to crash, this helper function just captures any code from that
DWORD SafeNtGetNextThread(HANDLE ProcessHandle, HANDLE ThreadHandle, ACCESS_MASK DesiredAccess, ULONG HandleAttributes,
    ULONG Flags, PHANDLE NewThreadHandle) {
    __try {
        return NtGetNextThread(ProcessHandle, ThreadHandle, DesiredAccess, HandleAttributes, Flags, NewThreadHandle);
    } __except (EXCEPTION_EXECUTE_HANDLER) {
        return GetExceptionCode();
    }
}
```
(couldn't put that into ThreadFreezer ctor since VS would complain with `error C2712: Cannot use __try in functions that require object unwinding`, works fine as a helper func though)

Wonder why this exception is only happening on UWP games though, could UWP maybe be turning something on that would make it throw exceptions instead of just returning status code?

If that can be disabled during the ThreadFreezer code it might be better than polluting the lib with SEH stuff.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -429,6 +429,23 @@ if(SAFETYHOOK_BUILD_TESTS) # build-tests
 	endif()
 
 endif()
+# Target: test4
+if(SAFETYHOOK_BUILD_TESTS) # build-tests
+	set(test4_SOURCES
+		"tests/test4.cpp"
+		cmake.toml
+	)
+
+	add_library(test4 SHARED)
+
+	target_sources(test4 PRIVATE ${test4_SOURCES})
+	source_group(TREE ${CMAKE_CURRENT_SOURCE_DIR} FILES ${test4_SOURCES})
+
+	target_link_libraries(test4 PUBLIC
+		safetyhook::safetyhook
+	)
+
+endif()
 # Target: unittest
 if(SAFETYHOOK_BUILD_TESTS) # build-tests
 	set(unittest_SOURCES

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -441,6 +441,42 @@ if(SAFETYHOOK_BUILD_TESTS) # build-tests
 	target_sources(test4 PRIVATE ${test4_SOURCES})
 	source_group(TREE ${CMAKE_CURRENT_SOURCE_DIR} FILES ${test4_SOURCES})
 
+	target_compile_features(test4 PUBLIC
+		cxx_std_23
+	)
+
+	if(MSVC) # msvc
+		target_compile_options(test4 PRIVATE
+			"/WX"
+			"/permissive-"
+			"/W4"
+			"/w14640"
+			"/EHsc"
+		)
+	endif()
+
+	if((CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND NOT CMAKE_CXX_COMPILER_FRONTEND_VARIANT MATCHES "^MSVC$") OR (CMAKE_C_COMPILER_ID MATCHES "Clang" AND NOT CMAKE_C_COMPILER_FRONTEND_VARIANT MATCHES "^MSVC$")) # clang
+		target_compile_options(test4 PRIVATE
+			-Werror
+			-Wall
+			-Wextra
+			-Wshadow
+			-Wnon-virtual-dtor
+			-pedantic
+		)
+	endif()
+
+	if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID STREQUAL "GNU") # gcc
+		target_compile_options(test4 PRIVATE
+			-Werror
+			-Wall
+			-Wextra
+			-Wshadow
+			-Wnon-virtual-dtor
+			-pedantic
+		)
+	endif()
+
 	target_link_libraries(test4 PUBLIC
 		safetyhook::safetyhook
 	)

--- a/cmake.toml
+++ b/cmake.toml
@@ -120,6 +120,12 @@ sources = ["tests/test2.cpp"]
 type = "test"
 sources = ["tests/test3.cpp"]
 
+[target.test4]
+condition = "build-tests"
+type = "shared"
+sources = ["tests/test4.cpp"]
+link-libraries = ["safetyhook::safetyhook"]
+
 [target.unittest]
 type = "test"
 sources = ["unittest/*.cpp"]

--- a/cmake.toml
+++ b/cmake.toml
@@ -104,6 +104,15 @@ msvc.private-compile-options = ["/WX", "/permissive-", "/W4", "/w14640", "/EHsc"
 clang.private-compile-options = ["-Werror", "-Wall", "-Wextra", "-Wshadow", "-Wnon-virtual-dtor", "-pedantic"]
 gcc.private-compile-options = ["-Werror", "-Wall", "-Wextra", "-Wshadow", "-Wnon-virtual-dtor", "-pedantic"]
 
+[template.test-dll]
+condition = "build-tests"
+type = "shared"
+link-libraries = ["safetyhook::safetyhook"]
+compile-features = ["cxx_std_23"]
+msvc.private-compile-options = ["/WX", "/permissive-", "/W4", "/w14640", "/EHsc"]
+clang.private-compile-options = ["-Werror", "-Wall", "-Wextra", "-Wshadow", "-Wnon-virtual-dtor", "-pedantic"]
+gcc.private-compile-options = ["-Werror", "-Wall", "-Wextra", "-Wshadow", "-Wnon-virtual-dtor", "-pedantic"]
+
 [target.test0]
 type = "test"
 sources = ["tests/test0.cpp"]
@@ -121,10 +130,8 @@ type = "test"
 sources = ["tests/test3.cpp"]
 
 [target.test4]
-condition = "build-tests"
-type = "shared"
+type = "test-dll"
 sources = ["tests/test4.cpp"]
-link-libraries = ["safetyhook::safetyhook"]
 
 [target.unittest]
 type = "test"

--- a/include/safetyhook.hpp
+++ b/include/safetyhook.hpp
@@ -10,4 +10,3 @@ using SafetyHookInline = safetyhook::InlineHook;
 using SafetyHookMid = safetyhook::MidHook;
 using SafetyInlineHook [[deprecated("Use SafetyHookInline instead.")]] = safetyhook::InlineHook;
 using SafetyMidHook [[deprecated("Use SafetyHookMid instead.")]] = safetyhook::MidHook;
-using SafetyThreadFreezer = safetyhook::ThreadFreezer;

--- a/include/safetyhook/allocator.hpp
+++ b/include/safetyhook/allocator.hpp
@@ -114,10 +114,10 @@ private:
         const std::vector<uintptr_t>& desired_addresses, size_t size, size_t max_distance = 0x7FFF'FFFF);
     void internal_free(uintptr_t address, size_t size);
 
-    void combine_adjacent_freenodes(Memory& memory);
-    [[nodiscard]] std::expected<uintptr_t, Error> allocate_nearby_memory(
+    static void combine_adjacent_freenodes(Memory& memory);
+    [[nodiscard]] static std::expected<uintptr_t, Error> allocate_nearby_memory(
         const std::vector<uintptr_t>& desired_addresses, size_t size, size_t max_distance);
-    [[nodiscard]] bool in_range(
+    [[nodiscard]] static bool in_range(
         uintptr_t address, const std::vector<uintptr_t>& desired_addresses, size_t max_distance);
 };
 } // namespace safetyhook

--- a/include/safetyhook/thread_freezer.hpp
+++ b/include/safetyhook/thread_freezer.hpp
@@ -4,51 +4,24 @@
 #pragma once
 
 #include <cstdint>
+#include <functional>
 #include <vector>
 
 #include <Windows.h>
 
 namespace safetyhook {
+/// @brief Executes a function while all other threads are frozen. Also allows for visiting each frozen thread and
+/// modifying it's context.
+/// @param run_fn The function to run while all other threads are frozen.
+/// @param visit_fn The function that will be called for each frozen thread.
+/// @note The visit function will be called in the order that the threads were frozen.
+/// @note The visit function will be called before the run function.
+/// @note Keep the logic inside run_fn and visit_fn as simple as possible to avoid deadlocks.
+void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint32_t, HANDLE, CONTEXT&)> visit_fn = {});
 
-/// @brief A class for freezing all threads in the process.
-class ThreadFreezer final {
-public:
-    ThreadFreezer(ThreadFreezer&) = delete;
-    ThreadFreezer(ThreadFreezer&&) noexcept = default;
-    ThreadFreezer& operator=(ThreadFreezer&) = delete;
-    ThreadFreezer& operator=(ThreadFreezer&&) noexcept = default;
-
-    /// @brief Constructs a new ThreadFreezer object.
-    /// @note This freezes all threads in the process except the current thread.
-    ThreadFreezer();
-
-    /// @brief Destructs the ThreadFreezer object.
-    /// @note This unfreezes all threads in the process.
-    ~ThreadFreezer();
-
-    /// @brief Fixes any threads that are currently on `old_ip` and sets them to `new_ip`.
-    /// @param old_ip The old instruction pointer.
-    /// @param new_ip The new instruction pointer.
-    void fix_ip(uintptr_t old_ip, uintptr_t new_ip);
-
-private:
-    struct FrozenThread {
-        uint32_t thread_id{};
-        HANDLE handle{};
-        CONTEXT ctx{};
-    };
-
-    // Custom allocator for `m_frozen_threads` that will work while threads are being frozen. This is necessary because
-    // `new` can deadlock while freezing threads.
-    template <class T> struct HeapAllocator {
-        using value_type = T;
-
-        HeapAllocator() = default;
-        template <class U> constexpr HeapAllocator(const HeapAllocator<U>&) noexcept {}
-        [[nodiscard]] T* allocate(size_t n) { return static_cast<T*>(HeapAlloc(GetProcessHeap(), 0, n * sizeof(T))); }
-        void deallocate(T* p, size_t) noexcept { HeapFree(GetProcessHeap(), 0, p); }
-    };
-
-    std::vector<FrozenThread, HeapAllocator<FrozenThread>> m_frozen_threads{};
-};
+/// @brief Will modify the context of a thread's IP to point to a new address if its IP is at the old address.
+/// @param ctx The thread context to modify.
+/// @param old_ip The old IP address.
+/// @param new_ip The new IP address.
+void fix_ip(CONTEXT& ctx, uintptr_t old_ip, uintptr_t new_ip);
 } // namespace safetyhook

--- a/include/safetyhook/thread_freezer.hpp
+++ b/include/safetyhook/thread_freezer.hpp
@@ -17,7 +17,8 @@ namespace safetyhook {
 /// @note The visit function will be called in the order that the threads were frozen.
 /// @note The visit function will be called before the run function.
 /// @note Keep the logic inside run_fn and visit_fn as simple as possible to avoid deadlocks.
-void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint32_t, HANDLE, CONTEXT&)> visit_fn = {});
+void execute_while_frozen(
+    const std::function<void()>& run_fn, const std::function<void(uint32_t, HANDLE, CONTEXT&)>& visit_fn = {});
 
 /// @brief Will modify the context of a thread's IP to point to a new address if its IP is at the old address.
 /// @param ctx The thread context to modify.

--- a/include/safetyhook/thread_freezer.hpp
+++ b/include/safetyhook/thread_freezer.hpp
@@ -38,6 +38,17 @@ private:
         CONTEXT ctx{};
     };
 
-    std::vector<FrozenThread> m_frozen_threads{};
+    // Custom allocator for `m_frozen_threads` that will work while threads are being frozen. This is necessary because
+    // `new` can deadlock while freezing threads.
+    template <class T> struct HeapAllocator {
+        using value_type = T;
+
+        HeapAllocator() = default;
+        template <class U> constexpr HeapAllocator(const HeapAllocator<U>&) noexcept {}
+        [[nodiscard]] T* allocate(size_t n) { return static_cast<T*>(HeapAlloc(GetProcessHeap(), 0, n * sizeof(T))); }
+        void deallocate(T* p, size_t) noexcept { HeapFree(GetProcessHeap(), 0, p); }
+    };
+
+    std::vector<FrozenThread, HeapAllocator<FrozenThread>> m_frozen_threads{};
 };
 } // namespace safetyhook

--- a/src/allocator.cpp
+++ b/src/allocator.cpp
@@ -1,5 +1,4 @@
 #include <algorithm>
-#include <cassert>
 #include <functional>
 #include <limits>
 
@@ -264,14 +263,10 @@ std::expected<uintptr_t, Allocator::Error> Allocator::allocate_nearby_memory(
 }
 
 bool Allocator::in_range(uintptr_t address, const std::vector<uintptr_t>& desired_addresses, size_t max_distance) {
-    for (auto&& desired_address : desired_addresses) {
+    return std::ranges::all_of(desired_addresses, [&](const auto& desired_address) {
         auto delta = (address > desired_address) ? address - desired_address : desired_address - address;
-        if (delta > max_distance) {
-            return false;
-        }
-    }
-
-    return true;
+        return delta <= max_distance;
+    });
 }
 
 Allocator::Memory::~Memory() {

--- a/src/inline_hook.cpp
+++ b/src/inline_hook.cpp
@@ -110,7 +110,7 @@ static void emit_jmp_e9(uintptr_t src, uintptr_t dst, size_t size = sizeof(JmpE9
 
 static bool decode(ZydisDecodedInstruction* ix, uintptr_t ip) {
     ZydisDecoder decoder{};
-    ZyanStatus status{};
+    ZyanStatus status;
 
 #if defined(_M_X64)
     status = ZydisDecoderInit(&decoder, ZYDIS_MACHINE_MODE_LONG_64, ZYDIS_STACK_WIDTH_64);

--- a/src/inline_hook.cpp
+++ b/src/inline_hook.cpp
@@ -320,15 +320,17 @@ std::expected<void, InlineHook::Error> InlineHook::e9_hook(const std::shared_ptr
 #endif
 
     // jmp from original to trampoline.
-    ThreadFreezer freezer{};
-
-    src = m_target;
-    dst = reinterpret_cast<uintptr_t>(&trampoline_epilogue->jmp_to_destination);
-    emit_jmp_e9(src, dst, m_original_bytes.size());
-
-    for (size_t i = 0; i < m_original_bytes.size(); ++i) {
-        freezer.fix_ip(m_target + i, m_trampoline.address() + i);
-    }
+    execute_while_frozen(
+        [this, &trampoline_epilogue] {
+            const auto src = m_target;
+            const auto dst = reinterpret_cast<uintptr_t>(&trampoline_epilogue->jmp_to_destination);
+            emit_jmp_e9(src, dst, m_original_bytes.size());
+        },
+        [this](uint32_t, HANDLE, CONTEXT& ctx) {
+            for (size_t i = 0; i < m_original_bytes.size(); ++i) {
+                fix_ip(ctx, m_target + i, m_trampoline.address() + i);
+            }
+        });
 
     return {};
 }
@@ -376,16 +378,18 @@ std::expected<void, InlineHook::Error> InlineHook::ff_hook(const std::shared_ptr
     emit_jmp_ff(src, dst, data);
 
     // jmp from original to trampoline.
-    ThreadFreezer freezer{};
-
-    src = m_target;
-    dst = m_destination;
-    data = src + sizeof(JmpFF);
-    emit_jmp_ff(src, dst, data, m_original_bytes.size());
-
-    for (size_t i = 0; i < m_original_bytes.size(); ++i) {
-        freezer.fix_ip(m_target + i, m_trampoline.address() + i);
-    }
+    execute_while_frozen(
+        [this] {
+            const auto src = m_target;
+            const auto dst = m_destination;
+            const auto data = src + sizeof(JmpFF);
+            emit_jmp_ff(src, dst, data, m_original_bytes.size());
+        },
+        [this](uint32_t, HANDLE, CONTEXT& ctx) {
+            for (size_t i = 0; i < m_original_bytes.size(); ++i) {
+                fix_ip(ctx, m_target + i, m_trampoline.address() + i);
+            }
+        });
 
     return {};
 }
@@ -398,14 +402,16 @@ void InlineHook::destroy() {
         return;
     }
 
-    ThreadFreezer freezer{};
-    UnprotectMemory unprotect{m_target, m_original_bytes.size()};
-
-    std::copy(m_original_bytes.begin(), m_original_bytes.end(), reinterpret_cast<uint8_t*>(m_target));
-
-    for (size_t i = 0; i < m_trampoline_size; ++i) {
-        freezer.fix_ip(m_trampoline.address() + i, m_target + i);
-    }
+    execute_while_frozen(
+        [this] {
+            UnprotectMemory unprotect{m_target, m_original_bytes.size()};
+            std::copy(m_original_bytes.begin(), m_original_bytes.end(), reinterpret_cast<uint8_t*>(m_target));
+        },
+        [this](uint32_t, HANDLE, CONTEXT& ctx) {
+            for (size_t i = 0; i < m_original_bytes.size(); ++i) {
+                fix_ip(ctx, m_trampoline.address() + i, m_target + i);
+            }
+        });
 
     m_trampoline.free();
 }

--- a/src/thread_freezer.cpp
+++ b/src/thread_freezer.cpp
@@ -16,6 +16,8 @@ NtGetNextThread(HANDLE ProcessHandle, HANDLE ThreadHandle, ACCESS_MASK DesiredAc
 
 namespace safetyhook {
 ThreadFreezer::ThreadFreezer() {
+    m_frozen_threads.reserve(1024);
+
     size_t num_threads_frozen{};
 
     do {

--- a/src/thread_freezer.cpp
+++ b/src/thread_freezer.cpp
@@ -23,13 +23,20 @@ ThreadFreezer::ThreadFreezer() {
         HANDLE thread{};
 
         while (true) {
+            HANDLE next_thread{};
             const auto status = NtGetNextThread(GetCurrentProcess(), thread,
                 THREAD_QUERY_LIMITED_INFORMATION | THREAD_SUSPEND_RESUME | THREAD_GET_CONTEXT | THREAD_SET_CONTEXT, 0,
-                0, &thread);
+                0, &next_thread);
+
+            if (thread != nullptr) {
+                CloseHandle(thread);
+            }
 
             if (!NT_SUCCESS(status)) {
                 break;
             }
+
+            thread = next_thread;
 
             const auto thread_id = GetThreadId(thread);
             const auto already_frozen = std::any_of(m_frozen_threads.begin(), m_frozen_threads.end(),
@@ -37,7 +44,6 @@ ThreadFreezer::ThreadFreezer() {
 
             // Don't freeze ourselves or threads we already froze.
             if (thread_id == 0 || thread_id == GetCurrentThreadId() || already_frozen) {
-                CloseHandle(thread);
                 continue;
             }
 
@@ -46,11 +52,11 @@ ThreadFreezer::ThreadFreezer() {
             thread_ctx.ContextFlags = CONTEXT_FULL;
 
             if (SuspendThread(thread) == static_cast<DWORD>(-1) || GetThreadContext(thread, &thread_ctx) == FALSE) {
-                CloseHandle(thread);
                 continue;
             }
 
             m_frozen_threads.push_back({thread_id, thread, thread_ctx});
+            thread = nullptr;
         }
     } while (num_threads_frozen != m_frozen_threads.size());
 }

--- a/src/thread_freezer.cpp
+++ b/src/thread_freezer.cpp
@@ -18,6 +18,7 @@ namespace safetyhook {
 void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint32_t, HANDLE, CONTEXT&)> visit_fn) {
     // Freeze all threads.
     int num_threads_frozen{};
+    auto first_run = true;
 
     do {
         HANDLE thread{};
@@ -51,7 +52,9 @@ void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint3
                 continue;
             }
 
-            if (suspend_count != 0) { // Thread was already suspended.
+            // Check if the thread was already frozen. Only resume if the thread was already frozen and it wasn't the
+            // first run of this freeze loop to account for threads that may have already been frozen for other reasons.
+            if (suspend_count != 0 && !first_run) {
                 ResumeThread(thread);
                 continue;
             }
@@ -70,6 +73,8 @@ void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint3
 
             ++num_threads_frozen;
         }
+
+        first_run = false;
     } while (num_threads_frozen != 0);
 
     // Run the function.

--- a/src/thread_freezer.cpp
+++ b/src/thread_freezer.cpp
@@ -1,5 +1,3 @@
-#include <algorithm>
-
 #include <Windows.h>
 #include <winternl.h>
 
@@ -15,14 +13,15 @@ NtGetNextThread(HANDLE ProcessHandle, HANDLE ThreadHandle, ACCESS_MASK DesiredAc
 }
 
 namespace safetyhook {
-void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint32_t, HANDLE, CONTEXT&)> visit_fn) {
+void execute_while_frozen(
+    const std::function<void()>& run_fn, const std::function<void(uint32_t, HANDLE, CONTEXT&)>& visit_fn) {
     // Freeze all threads.
-    int num_threads_frozen{};
+    int num_threads_frozen;
     auto first_run = true;
 
     do {
-        HANDLE thread{};
         num_threads_frozen = 0;
+        HANDLE thread{};
 
         while (true) {
             HANDLE next_thread{};
@@ -52,14 +51,14 @@ void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint3
                 continue;
             }
 
-            // Check if the thread was already frozen. Only resume if the thread was already frozen and it wasn't the
+            // Check if the thread was already frozen. Only resume if the thread was already frozen, and it wasn't the
             // first run of this freeze loop to account for threads that may have already been frozen for other reasons.
             if (suspend_count != 0 && !first_run) {
                 ResumeThread(thread);
                 continue;
             }
 
-            auto thread_ctx = CONTEXT{};
+            CONTEXT thread_ctx{};
 
             thread_ctx.ContextFlags = CONTEXT_FULL;
 

--- a/src/thread_freezer.cpp
+++ b/src/thread_freezer.cpp
@@ -15,12 +15,13 @@ NtGetNextThread(HANDLE ProcessHandle, HANDLE ThreadHandle, ACCESS_MASK DesiredAc
 }
 
 namespace safetyhook {
-ThreadFreezer::ThreadFreezer() {
-    size_t num_threads_frozen{};
+void execute_while_frozen(std::function<void()> run_fn, std::function<void(uint32_t, HANDLE, CONTEXT&)> visit_fn) {
+    // Freeze all threads.
+    int num_threads_frozen{};
 
     do {
-        num_threads_frozen = m_frozen_threads.size();
         HANDLE thread{};
+        num_threads_frozen = 0;
 
         while (true) {
             HANDLE next_thread{};
@@ -39,11 +40,19 @@ ThreadFreezer::ThreadFreezer() {
             thread = next_thread;
 
             const auto thread_id = GetThreadId(thread);
-            const auto already_frozen = std::any_of(m_frozen_threads.begin(), m_frozen_threads.end(),
-                [=](const auto& thread) { return thread.thread_id == thread_id; });
 
-            // Don't freeze ourselves or threads we already froze.
-            if (thread_id == 0 || thread_id == GetCurrentThreadId() || already_frozen) {
+            if (thread_id == 0 || thread_id == GetCurrentThreadId()) {
+                continue;
+            }
+
+            const auto suspend_count = SuspendThread(thread);
+
+            if (suspend_count == static_cast<DWORD>(-1)) {
+                continue;
+            }
+
+            if (suspend_count != 0) { // Thread was already suspended.
+                ResumeThread(thread);
                 continue;
             }
 
@@ -51,41 +60,67 @@ ThreadFreezer::ThreadFreezer() {
 
             thread_ctx.ContextFlags = CONTEXT_FULL;
 
-            if (SuspendThread(thread) == static_cast<DWORD>(-1) || GetThreadContext(thread, &thread_ctx) == FALSE) {
+            if (GetThreadContext(thread, &thread_ctx) == FALSE) {
                 continue;
             }
 
-            m_frozen_threads.push_back({thread_id, thread, thread_ctx});
-            thread = nullptr;
-        }
-    } while (num_threads_frozen != m_frozen_threads.size());
-}
+            if (visit_fn) {
+                visit_fn(thread_id, thread, thread_ctx);
+            }
 
-ThreadFreezer::~ThreadFreezer() {
-    for (auto& thread : m_frozen_threads) {
-        SetThreadContext(thread.handle, &thread.ctx);
-        ResumeThread(thread.handle);
-        CloseHandle(thread.handle);
+            ++num_threads_frozen;
+        }
+    } while (num_threads_frozen != 0);
+
+    // Run the function.
+    if (run_fn) {
+        run_fn();
+    }
+
+    // Resume all threads.
+    HANDLE thread{};
+
+    while (true) {
+        HANDLE next_thread{};
+        const auto status = NtGetNextThread(GetCurrentProcess(), thread,
+            THREAD_QUERY_LIMITED_INFORMATION | THREAD_SUSPEND_RESUME | THREAD_GET_CONTEXT | THREAD_SET_CONTEXT, 0, 0,
+            &next_thread);
+
+        if (thread != nullptr) {
+            CloseHandle(thread);
+        }
+
+        if (!NT_SUCCESS(status)) {
+            break;
+        }
+
+        thread = next_thread;
+
+        const auto thread_id = GetThreadId(thread);
+
+        if (thread_id == 0 || thread_id == GetCurrentThreadId()) {
+            continue;
+        }
+
+        ResumeThread(thread);
     }
 }
 
-void ThreadFreezer::fix_ip(uintptr_t old_ip, uintptr_t new_ip) {
-    for (auto& thread : m_frozen_threads) {
+void fix_ip(CONTEXT& ctx, uintptr_t old_ip, uintptr_t new_ip) {
 #ifdef _M_X64
-        auto ip = thread.ctx.Rip;
+    auto ip = ctx.Rip;
 #else
-        auto ip = thread.ctx.Eip;
+    auto ip = ctx.Eip;
 #endif
 
-        if (ip == old_ip) {
-            ip = new_ip;
-        }
-
-#ifdef _M_X64
-        thread.ctx.Rip = ip;
-#else
-        thread.ctx.Eip = ip;
-#endif
+    if (ip == old_ip) {
+        ip = new_ip;
     }
+
+#ifdef _M_X64
+    ctx.Rip = ip;
+#else
+    ctx.Eip = ip;
+#endif
 }
 } // namespace safetyhook

--- a/src/thread_freezer.cpp
+++ b/src/thread_freezer.cpp
@@ -16,8 +16,6 @@ NtGetNextThread(HANDLE ProcessHandle, HANDLE ThreadHandle, ACCESS_MASK DesiredAc
 
 namespace safetyhook {
 ThreadFreezer::ThreadFreezer() {
-    m_frozen_threads.reserve(1024);
-
     size_t num_threads_frozen{};
 
     do {

--- a/tests/test4.cpp
+++ b/tests/test4.cpp
@@ -5,25 +5,23 @@
 SafetyHookInline g_PeekMessageA_hook{};
 SafetyHookInline g_PeekMessageW_hook{};
 
-BOOL WINAPI hooked_PeekMessageA(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilterMax, UINT wRemoveMsg)
-{
+BOOL WINAPI hooked_PeekMessageA(LPMSG msg, HWND wnd, UINT filter_min, UINT filter_max, UINT remove_msg) {
     OutputDebugString("hooked_PeekMessageA\n");
-    return g_PeekMessageA_hook.stdcall<BOOL>(lpMsg, hWnd, wMsgFilterMin, wMsgFilterMax, wRemoveMsg);
+    return g_PeekMessageA_hook.stdcall<BOOL>(msg, wnd, filter_min, filter_max, remove_msg);
 }
 
-BOOL WINAPI hooked_PeekMessageW(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilterMax, UINT wRemoveMsg)
-{
+BOOL WINAPI hooked_PeekMessageW(LPMSG msg, HWND wnd, UINT filter_min, UINT filter_max, UINT remove_msg) {
     OutputDebugString("hooked_PeekMessageW\n");
-    return g_PeekMessageW_hook.stdcall<BOOL>(lpMsg, hWnd, wMsgFilterMin, wMsgFilterMax, wRemoveMsg);
+    return g_PeekMessageW_hook.stdcall<BOOL>(msg, wnd, filter_min, filter_max, remove_msg);
 }
 
-BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
-{
-    switch (fdwReason)
-    {
+BOOL WINAPI DllMain(HINSTANCE, DWORD reason, LPVOID) {
+    switch (reason) {
     case DLL_PROCESS_ATTACH:
-        g_PeekMessageA_hook = safetyhook::create_inline(&PeekMessageA, &hooked_PeekMessageA);
-        g_PeekMessageW_hook = safetyhook::create_inline(&PeekMessageW, &hooked_PeekMessageW);
+        g_PeekMessageA_hook = safetyhook::create_inline(
+            reinterpret_cast<void*>(&PeekMessageA), reinterpret_cast<void*>(&hooked_PeekMessageA));
+        g_PeekMessageW_hook = safetyhook::create_inline(
+            reinterpret_cast<void*>(&PeekMessageW), reinterpret_cast<void*>(&hooked_PeekMessageW));
         break;
 
     case DLL_PROCESS_DETACH:

--- a/tests/test4.cpp
+++ b/tests/test4.cpp
@@ -1,0 +1,35 @@
+#include <Windows.h>
+
+#include <safetyhook.hpp>
+
+SafetyHookInline g_PeekMessageA_hook{};
+SafetyHookInline g_PeekMessageW_hook{};
+
+BOOL WINAPI hooked_PeekMessageA(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilterMax, UINT wRemoveMsg)
+{
+    OutputDebugString("hooked_PeekMessageA\n");
+    return g_PeekMessageA_hook.stdcall<BOOL>(lpMsg, hWnd, wMsgFilterMin, wMsgFilterMax, wRemoveMsg);
+}
+
+BOOL WINAPI hooked_PeekMessageW(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilterMax, UINT wRemoveMsg)
+{
+    OutputDebugString("hooked_PeekMessageW\n");
+    return g_PeekMessageW_hook.stdcall<BOOL>(lpMsg, hWnd, wMsgFilterMin, wMsgFilterMax, wRemoveMsg);
+}
+
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+{
+    switch (fdwReason)
+    {
+    case DLL_PROCESS_ATTACH:
+        g_PeekMessageA_hook = safetyhook::create_inline(&PeekMessageA, &hooked_PeekMessageA);
+        g_PeekMessageW_hook = safetyhook::create_inline(&PeekMessageW, &hooked_PeekMessageW);
+        break;
+
+    case DLL_PROCESS_DETACH:
+        g_PeekMessageW_hook.reset();
+        g_PeekMessageA_hook.reset();
+        break;
+    }
+    return TRUE;
+}

